src/v/cluster/distributed_kv_stm.h

```diff
@@ -0,0 +1,422 @@
+      , _default_max_partitions(max_partitions)
+      , _is_routing_partition(_raft->ntp().tp.partition == routing_partition){};
+
+    ss::future<> start() override { co_await persisted_stm<>::start(); }
+    ss::future<> stop() override { co_await _gate.close(); }
+
+    ss::future<> apply(const model::record_batch& record_batch) override {
+        if (record_batch.header().type != model::record_batch_type::raft_data) {
+            co_return;
+        }
+        auto holder = _gate.hold();
+        auto units = co_await _lock.hold_read_lock();
+        co_await record_batch.for_each_record_async(
+          [this](model::record r) -> ss::future<> {
+              auto op = reflection::from_iobuf<operation>(r.release_value());
+              if (op.type == op_type::repartitioning) {
+                  do_apply_repartition(std::move(op.data));
+              } else if (op.type == op_type::coordinator_assignment) {
+                  do_apply_assignment(std::move(op.data));
+              } else if (op.type == op_type::kv_data) {
+                  do_apply_offsets(std::move(op.data));
+              } else {
+                  vlog(clusterlog.error, "unknown type: {}", (int)op.type);
+              }
+              return ss::now();
+          });
+    }
+
+    ss::future<>
+    apply_local_snapshot(stm_snapshot_header header, iobuf&& bytes) override {
+        auto holder = _gate.hold();
+        auto units = _lock.hold_write_lock();
+
+        auto snapshot_offset = header.offset;
+        iobuf_parser parser(std::move(bytes));
+        auto snap = co_await serde::read_async<snapshot>(parser);
+
+        if (_is_routing_partition) {
+            _num_partitions = snap.num_partitions;
+            _coordinators.clear();
+            if (snap.coordinators) {
+                _coordinators = std::move(snap.coordinators.value());
+            }
+        }
+        kv_data = std::move(snap.kv_data);
+        set_next(model::next_offset(snapshot_offset));
+    }
+
+    ss::future<stm_snapshot> take_local_snapshot() override {
+        auto holder = _gate.hold();
+        auto units = _lock.hold_write_lock();
+        auto last_applied = last_applied_offset();
+        snapshot result;
+        if (_is_routing_partition) {
+            result.num_partitions = result.num_partitions;
+            result.coordinators = _coordinators;
+        }
+        result.kv_data = kv_data;
+        iobuf result_buf;
+        co_await serde::write_async(result_buf, std::move(result));
+        co_return stm_snapshot::create(0, last_applied, std::move(result_buf));
+    }
+
+    ss::future<> apply_raft_snapshot(const iobuf&) final {
+        // kv expiry is currently explicit using forget(), nothing to
+        // do on log truncations.
+        co_return;
+    }
+
+    std::string_view get_name() const final { return "kv_stm"; }
+    ss::future<iobuf> take_snapshot(model::offset) final { co_return iobuf{}; }
```

Comment:
What is the story behind keeping the underlying partition size controlled ? What retention policy are we going to apply ?

Reply:
For the incremental snapshot we may always read log and apply it to the previous snapshot up to the given offset so no additional state is required

Reply:
chatted offline. added a todo for future.